File: 算法基础课(修订版)/笔试拿分技巧/P14162/verify_acm.cc
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int solve(int n, int k, vector<pair<int, int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        int start = 0;
        priority_queue<int, vector<int>, greater<int>> heap;
        int idx = 0;
        int ans = 0;

        while (idx < n || !heap.empty()) {
            while (!heap.empty() && heap.top() < start) {
                heap.pop();
            }

            while (idx < n && intervals[idx].first <= start) {
                heap.push(intervals[idx].second);
                idx++;
            }

            if (heap.empty()) {
                if (idx == n) {
                    break;
                }
                start = max(start, intervals[idx].first);
            }

            while (idx < n && intervals[idx].first <= start) {
                heap.push(intervals[idx].second);
                idx++;
            }

            for (int i = 0; i < k; ++i) {
                if (!heap.empty()) {
                    ans++;
                    heap.pop();
                }
            }

            start++;
        }

        return ans;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, k;
    cin >> n >> k;
    vector<pair<int, int>> intervals(n);
    for (int i = 0; i < n; ++i) {
        cin >> intervals[i].first >> intervals[i].second;
    }

    Solution solution;
    cout << solution.solve(n, k, intervals);
    return 0;
}

File: 算法基础课(修订版)/笔试拿分技巧/P14162/verify_acm.py
import heapq


class Solution:
    def solve(self, n, k, intervals):
        intervals.sort(key=lambda x: x[0])
        start = 0
        heap = []
        idx = 0
        ans = 0

        while idx < n or heap:
            while heap and heap[0] < start:
                heapq.heappop(heap)

            while idx < n and intervals[idx][0] <= start:
                heapq.heappush(heap, intervals[idx][1])
                idx += 1

            if not heap:
                if idx == n:
                    break
                start = max(start, intervals[idx][0])

            while idx < n and intervals[idx][0] <= start:
                heapq.heappush(heap, intervals[idx][1])
                idx += 1

            for _ in range(k):
                if heap:
                    ans += 1
                    heapq.heappop(heap)

            start += 1

        return ans


if __name__ == "__main__":
    n, k = map(int, input().split())
    intervals = [list(map(int, input().split())) for _ in range(n)]

    solution = Solution()
    print(solution.solve(n, k, intervals))
